contracts/masset/liquidator/Liquidator.sol

```diff
@@ -0,0 +1,329 @@
+
+        // The minimum amount of output tokens that must be received for the transaction not to revert
+        // Set to 80% of expected
+        uint256 minAcceptable = expectedAmount.mul(uint(8000)).div(uint(10000));
+
+        // Sell amountToSell unless balance is lower in which case sell everything and relax acceptable check
+        uint256 sellAmount;
+        if (bal > amountToSell) {
+            sellAmount = amountToSell;
+        } else {
+            sellAmount = bal;
+            minAcceptable = 0;
+        }
+
+        // Approve Uniswap and make the swap
+        // https://uniswap.org/docs/v2/smart-contracts/router02/#swapexacttokensfortokens
+        IERC20(sellToken).safeApprove(uniswapAddress, 0);
+        IERC20(sellToken).safeApprove(uniswapAddress, amountToSell);
+        IUniswapV2Router02(uniswapAddress).swapExactTokensForTokens(
+            sellAmount,     
+            minAcceptable,
+            liquidation.uniswapPath,
+            address(this),
+            now.add(1800)
+        );
+
+        // Deposit to lending platform
+        // Assumes integration contracts have inifinte approval to collect them
+        if (liquidation.lendingPlatform == LendingPlatform.Compound) {
+            depositToCompound(liquidation.pToken, bAsset); 
+        } else {
+            revert("Lending Platform not supported");
+        }
+
+        emit LiquidationTriggered(_bAsset);
+    }
+
+    /**
+    * @dev Deposits to Compound 
+    * @param _pToken The _pToken to mint
+    * @param _bAsset The _bAsset liquidation to be triggered
+    */
+    function depositToCompound(address _pToken, address _bAsset)
+        internal
+    {
+        uint256 bAssetBalance = IERC20(_bAsset).balanceOf(address(this));
+        require((bAssetBalance > 0), "No tokens to deposit");
+        require(ICERC20(_pToken).mint(bAssetBalance) == 0, "cToken mint failed");
+    }
+
+    /**
+    * @dev Get the amount of sellToken to be sold for a number of bAsset
+    * @param _uniswapPath The Uniswap path for this liquidation
+    * @param _trancheAmount The tranche size that we want to buy each time 
+    */
+    function getAmountToSell(
+        address[]   memory _uniswapPath,
+        uint256      _trancheAmount
+    )
+        internal view returns (uint256, uint256)
+    {
+
+        // The _trancheAmount is the number of bAsset we want to buy each time
+        // DAI has 18 decimals so 1000 DAI is 10*10^18 or 1000000000000000000000
+        // This value is set when adding the liquidation
+        // We randomize this amount by buying betwen 80% and 95% of the amount.
+        // Uniswap gives us the amount we need to sell with `getAmountsIn`.
+        uint256 randomBasisPoint = uint256(blockhash(block.number-1)).mod(uint(1500)).add(uint(8000));
+        uint256 amountWanted = _trancheAmount.mul(randomBasisPoint).div(uint(10000));
```

Comment:
This is intended to randomise the tranche amount. So if `_trancheAmount` is 1000 DAI it will generate a number between 800 & 950 to send to Uniswap to find out how much COMP to sell. If we don't care about the same `_trancheAmount` each time these lines are obsolete.